pallets/liquidity-pools-gateway/queue/src/lib.rs

```diff
@@ -0,0 +1,276 @@
+	#[pallet::getter(fn failed_message_queue)]
+	pub type FailedMessageQueue<T: Config> =
+		StorageMap<_, Blake2_128Concat, T::MessageNonce, (T::Message, DispatchError)>;
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub (super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// A message was submitted.
+		MessageSubmitted {
+			nonce: T::MessageNonce,
+			message: T::Message,
+		},
+
+		/// Message execution success.
+		MessageExecutionSuccess {
+			nonce: T::MessageNonce,
+			message: T::Message,
+		},
+
+		/// Message execution failure.
+		MessageExecutionFailure {
+			nonce: T::MessageNonce,
+			message: T::Message,
+			error: DispatchError,
+		},
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// Message not found.
+		MessageNotFound,
+	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		fn on_idle(_now: BlockNumberFor<T>, max_weight: Weight) -> Weight {
+			Self::service_message_queue(max_weight)
+		}
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Convenience method for manually processing a message.
+		///
+		/// If the execution fails, the message gets moved to the
+		/// `FailedMessageQueue` storage.
+		///
+		/// NOTE - this extrinsic does not error out during message processing
+		/// to ensure that any storage changes (i.e. to the message queues)
+		/// are not reverted.
+		#[pallet::weight(T::WeightInfo::process_message())]
+		#[pallet::call_index(0)]
+		pub fn process_message(
+			origin: OriginFor<T>,
+			nonce: T::MessageNonce,
+		) -> DispatchResultWithPostInfo {
+			ensure_signed(origin)?;
+
+			let message = MessageQueue::<T>::take(nonce).ok_or(Error::<T>::MessageNotFound)?;
+
+			match Self::process_message_and_deposit_event(nonce, message.clone()) {
+				Ok(p) => Ok(p),
+				Err(e) => {
+					FailedMessageQueue::<T>::insert(nonce, (message, e.error));
+
+					Ok(e.post_info)
+				}
+			}
+		}
+
+		/// Convenience method for manually processing a failed message.
+		///
+		/// If the execution is successful, the message gets removed from the
+		/// `FailedMessageQueue` storage.
+		///
+		/// NOTE - this extrinsic does not error out during message processing
+		/// to ensure that any storage changes (i.e. to the message queues)
+		/// are not reverted.
+		#[pallet::weight(T::WeightInfo::process_failed_message())]
+		#[pallet::call_index(1)]
+		pub fn process_failed_message(
+			origin: OriginFor<T>,
+			nonce: T::MessageNonce,
+		) -> DispatchResultWithPostInfo {
+			ensure_signed(origin)?;
+
+			let (message, _) =
+				FailedMessageQueue::<T>::get(nonce).ok_or(Error::<T>::MessageNotFound)?;
+
+			match Self::process_message_and_deposit_event(nonce, message.clone()) {
+				Ok(p) => {
+					FailedMessageQueue::<T>::remove(nonce);
+
+					Ok(p)
+				}
+				Err(e) => Ok(e.post_info),
+			}
+		}
+	}
+
+	impl<T: Config> Pallet<T> {
+		fn process_message_and_deposit_event(
+			nonce: T::MessageNonce,
+			message: T::Message,
+		) -> DispatchResultWithPostInfo {
```

Comment:
Just one consideration about `DispatchResultWithPostInfo`. If possible, it's better not to use it as a way of measure the extrinsic. It should only be used for cases where you can optimize the extrinsic. It means only when the consumed weight is less than the expected weight. The reason is that the user (and the node) wants to know the fee cost of extrinsic costs upfront, to know if they have enough funds or if the extrinsic fit into the block.

In which cases, getting the weight from `T::MessageProcessor::process()` is reducing the weight from the original benchmarked weight?

Reply:
> In which cases, getting the weight from T::MessageProcessor::process() is reducing the weight from the original benchmarked weight?

We would have to implement the `MessageProcessor` trait in the LP gateway and figure out a way to properly test that then.

But coming back to your initial point - `process_message_and_deposit_event` should return `DispatchResultWithPostInfo` since it's used in the hook. I would drop it from the extrinsics though.